File: archive/ruby/muse/algorithms/number_theory/Factorial.rb
module Factorial
  def self.iterativeProcedure(n)
    return 0 if n.negative?

    result = 1
    1.upto(n) do |i|
      result *= i
    end

    result
  end

  def self.recursiveProcedure(n)
    return 0 if n.negative?

    return 1 if n.zero?

    recursiveProcedure(n - 1) * n
  end
end

File: archive/nodejs/muse/algorithms/number_theory/Factorial.js
export const iterativeProcedure = (n) => {
  if (n < 0) {
    return 0;
  }

  let result = 1;
  for (let i = 1; i <= n; i++) {
    result *= i;
  }

  return result;
};

export const recursiveProcedure = (n) => {
  if (n < 0) {
    return 0;
  }

  if (n === 0) {
    return 1;
  }
  return recursiveProcedure(n - 1) * n;
};
